jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/client/ClientUpgradeRequest.java

```diff
@@ -298,10 +298,13 @@ public void upgrade(HttpResponse response, HttpConnectionOverHTTP httpConnection
             }
         }
 
-        // Put requested internal Extensions into the negotiatedExtension list.
+        // Insert client requested internal Extensions into the negotiatedExtension list from the server.
+        // The order of extensions in the response is assumed to be the same as what the client negotiated.
         int i = 0;
         for (ExtensionConfig reqConfig : requestedExtensions)
         {
+            // If this is an internal extension insert in negotiatedExtensions at position i.
+            // If this is not an internal extension we must advance i only if this extension was negotiated by the server.
             if (reqConfig.isInternalExtension())
                 negotiatedExtensions.add(i++, reqConfig);
             else if (i < negotiatedExtensions.size() && reqConfig.getName().equals(negotiatedExtensions.get(i).getName()))
```

Comment:
This is really confusing.

What you want is to remove the extensions that the server did not negotiate from the list of the client extensions (internal and non-internal).

If that's correct, then I think this is more understandable (pseudo-code):

```java
for (Iterator<ExtensionConfig> iterator = clientExtensions.iterator(); iterator.hasNext();) {
  if (!extConfig.isInternal() && !serverExtensions.contains(extConfig)) {
    iterator.remove();
  }
}
```